lib/queue.js

```diff
@@ -41,21 +30,79 @@ function compare (a, b) {
     return matches;
 }
 
+/**
+ * Creates a new queue
+ * @class
+ */
 var FetchQueue = function() {
-    this.oldestUnfetchedIndex = 0;
-    this.completeCache = 0;
-    this.scanIndex = {};
+    Array.call(this);
+
+    /**
+     * Speeds up {@link FetchQueue.oldestUnfetchedItem} by storing the index at
+     * which the latest oldest unfetched queue item was found.
+     * @name FetchQueue._oldestUnfetchedIndex
+     * @private
+     * @type {Number}
+     */
+    Object.defineProperty(this, "_oldestUnfetchedIndex", {
+        enumerable: false,
+        writable: true,
+        value: 0
+    });
+
+    /**
+     * Serves as a cache for what URL's have been fetched. Keys are URL's,
+     * values are booleans.
+     * @name FetchQueue._scanIndex
+     * @private
+     * @type {Object}
+     */
+    Object.defineProperty(this, "_scanIndex", {
+        enumerable: false,
+        writable: true,
+        value: {}
+    });
+
+    /**
+     * Controls what properties can be operated on with the
+     * {@link FetchQueue#min}, {@link FetchQueue#avg} and {@link FetchQueue#max}
+     * methods.
+     * @name FetchQueue._allowedStatistics
+     * @type {Array}
+     */
+    Object.defineProperty(this, "_allowedStatistics", {
+        enumerable: false,
+        writable: true,
+        value: [
+            "actualDataSize",
+            "contentLength",
+            "downloadTime",
+            "requestLatency",
+            "requestTime"
+        ]
+    });
 };
 
-module.exports = FetchQueue;
+FetchQueue.prototype = Array.prototype;
 
-FetchQueue.prototype = [];
+/**
+ * Called when {@link FetchQueue#add} returns a result
+ * @callback FetchQueue~addCallback
+ * @param {Error} [error]         If the operation was successful, this will be `null`. Otherwise it will be the error that was encountered.
+ * @param {QueueItem} [queueItem] The queue item that was added to the queue. It's status property will have changed to `"queued"`.
+ */
+
+/**
+ * Adds an item to the queue
+ * @param {QueueItem} queueItem             Queue item that is to be added to the queue
+ * @param {Boolean} [force=false]           If true, the queue item will be added regardless of whether it already exists in the queue
+ * @param {FetchQueue~addCallback} callback
+ */
 FetchQueue.prototype.add = function(queueItem, force, callback) {
-    callback = callback instanceof Function ? callback : noop;
```

Comment:
In the current version of the queue, callbacks aren't optional, so I've removed all of these lines